File: app/services/property_web_scraper/url_validator.rb
module PropertyWebScraper
  # Validates and parses a URL for use with the scraper.
  #
  # Consolidates URL validation logic shared across controllers and
  # services: presence check, whitespace stripping, URI parsing,
  # HTTP(S) scheme check, and ImportHost lookup.
  #
  # @example
  #   result = UrlValidator.call("https://www.idealista.com/inmueble/123/")
  #   result.valid?       #=> true
  #   result.uri          #=> #<URI::HTTPS>
  #   result.import_host  #=> ImportHost
  class UrlValidator
    MISSING  = :missing
    INVALID  = :invalid
    UNSUPPORTED = :unsupported

    Result = Struct.new(:valid, :uri, :import_host, :error_message, :error_code, keyword_init: true) do
      alias_method :valid?, :valid
    end

    # Validates the given URL string.
    #
    # @param url [String, nil] the URL to validate
    # @return [Result]
    def self.call(url)
      if url.blank?
        return Result.new(valid: false, error_code: MISSING, error_message: "Please provide a url")
      end

      stripped = url.strip
      uri = begin
        URI.parse(stripped)
      rescue URI::InvalidURIError
        nil
      end

      unless uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS)
        return Result.new(valid: false, error_code: INVALID, error_message: "Please provide a valid url")
      end

      import_host = ImportHost.find_by_host(uri.host)
      unless import_host
        return Result.new(valid: false, error_code: UNSUPPORTED, error_message: "Sorry, the url provided is currently not supported")
      end

      Result.new(valid: true, uri: uri, import_host: import_host)
    end
  end
end

File: astro-app/src/lib/services/url-validator.ts
import { ImportHost } from '../models/import-host.js';

export interface ValidationResult {
  valid: boolean;
  uri?: URL;
  importHost?: ImportHost;
  errorMessage?: string;
  errorCode?: string;
}

export const MISSING = 'missing';
export const INVALID = 'invalid';
export const UNSUPPORTED = 'unsupported';

/**
 * Validate and parse a URL for use with the scraper.
 * Port of Ruby UrlValidator.call.
 */
export async function validateUrl(url: string | undefined | null): Promise<ValidationResult> {
  if (!url || url.trim() === '') {
    return { valid: false, errorCode: MISSING, errorMessage: 'Please provide a url' };
  }

  const stripped = url.trim();
  let uri: URL;
  try {
    uri = new URL(stripped);
  } catch {
    return { valid: false, errorCode: INVALID, errorMessage: 'Please provide a valid url' };
  }

  if (uri.protocol !== 'http:' && uri.protocol !== 'https:') {
    return { valid: false, errorCode: INVALID, errorMessage: 'Please provide a valid url' };
  }

  const importHost = await ImportHost.findByHost(uri.hostname);
  if (!importHost) {
    return {
      valid: false,
      errorCode: UNSUPPORTED,
      errorMessage: 'Sorry, the url provided is currently not supported',
    };
  }

  return { valid: true, uri, importHost };
}
